src/Microsoft.Health.Dicom.Api/Web/LazyMultipartReadOnlyStream.cs

```diff
@@ -0,0 +1,228 @@
+                    _currentStreamContent = _asyncEnumerator.Current;
+                    CopyStringToBuffer(GetHeaderBoundary(true));
+                }
+                else
+                {
+                    throw new InvalidOperationException("Enumerator cannot be empty");
+                }
+            }
+
+            if (_bufferPosition == _bufferLength && _currentStreamContent.Stream.Position == _currentStreamContent.Stream.Length)
+            {
+                if (await _asyncEnumerator.MoveNextAsync())
+                {
+                    await _currentStreamContent.Stream.DisposeAsync();
+                    _currentStreamContent = _asyncEnumerator.Current;
+                    CopyStringToBuffer(GetHeaderBoundary(false));
+                }
+                else
+                {
+                    await _currentStreamContent.Stream.DisposeAsync();
+                    CopyStringToBuffer(GetTerminatingBoundary());
+                    _terminating = true;
+                }
+            }
+
+            // get
+            if (!_terminating && _bufferPosition == _bufferLength)
+            {
+#pragma warning disable CA1835 // Prefer the 'Memory'-based overloads for 'ReadAsync' and 'WriteAsync', (Azure.Storage.LazyLoadingReadOnlyStream does not implement that.)
+                _bufferLength = await _currentStreamContent.Stream.ReadAsync(_buffer, 0, _bufferSize, cancellationToken);
+#pragma warning restore CA1835 // Prefer the 'Memory'-based overloads for 'ReadAsync' and 'WriteAsync'
+                _bufferPosition = 0;
+            }
+
+            // We will return the minimum of remainingBytesInBuffer and the count provided by the user
+            int remainingBytesInBuffer = _bufferLength - _bufferPosition;
+            int bytesToWrite = Math.Min(remainingBytesInBuffer, count);
+
+            // copy
+            Array.Copy(_buffer, _bufferPosition, buffer, offset, bytesToWrite);
+            _bufferPosition += bytesToWrite;
+
+            return bytesToWrite;
+        }
+
+        private int CopyStringToBuffer(string multipartHeader)
+        {
+            byte[] startBoundary = EncodeStringToByteArray(multipartHeader);
+            int bytesToWrite = startBoundary.Length;
+            Array.Copy(startBoundary, 0, _buffer, 0, bytesToWrite);
+            _bufferPosition = 0;
+            _bufferLength = bytesToWrite;
+
+            return bytesToWrite;
+        }
+
+        protected override void Dispose(bool disposing)
+        {
+            base.Dispose(disposing);
+            // Return the buffer to the pool if we're called from Dispose or a finalizer
+            if (_buffer != null)
+            {
+                ArrayPool<byte>.Shared.Return(_buffer, clearArray: true);
+                _buffer = null;
+            }
+        }
+
+        #region MultiPart boundary write
+        private string GetHeaderBoundary(bool isStart)
+        {
+            var multiPartStringBuilder = new StringBuilder();
```

Comment:
Would it make sense to set the capacity since we're essentially building the same string builder multiple times? Just wondering on allocations since the initial size seems to be 16 characters and we'll always overrun that and this could be called many times on a large request.

Reply:
Also, do we have to have a different start boundary from an intermediate boundary or can they both just be the same as the intermediate boundary?

Reply:
Start and intermediate have different formats

Reply:
I am not sure if we need to optimize StringBuilder, it is not immutable like string, so no copying around, and creates additional char[]s to allows easy updates.